File: objc/AbstractFactory/AbstractFactory/GUIBuilder.m
#import "GUIBuilder.h"

@implementation GUIBuilder {
    NSString* platformName;
    id<AbstractGUIFactory> guiFactory;
}

-(instancetype)initWithPlatform:(NSString *)platform {
    if(self = [super init]) {
        platformName = platform;
    }
    return self;
}

-(void)initGUIFactory {
    if(nil != guiFactory) return;
    if([platformName isEqualToString:@"Windows"]) guiFactory = [[WinFactory alloc] init];
    else guiFactory = [[OSXFactory alloc] init];
}

-(id<Window>)buildWindow {
    [self initGUIFactory];
    return [guiFactory createWindow];
}

-(id<Button>)buildButton {
    [self initGUIFactory];
    return [guiFactory createButton];
}

@end

File: swift/AbstractFactory/AbstractFactory/GUIBuilder.swift
import Foundation

class GUIBuilder {
    private var platform: String
    private var guiFactory: AbstractGUIFactory?
    
    init(platform: String) {
        self.platform = platform
    }
    
    func initGuiFactory() -> Void {
        if nil != guiFactory { return }
        if platform == "Windows" { guiFactory = WinFactory() }
        else { guiFactory = OSXFactory() }
    }
    
    func buildButton() -> Button {
        initGuiFactory()
        return guiFactory!.createButton()
    }
    
    func buildWindow() -> Window {
        initGuiFactory()
        return guiFactory!.createWindow()
    }

}
